File: Python/Array/309_Best Time to Buy and Sell Stock with Cooldown.py
################ Memoziation ##########################
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        
        dp = {}
        
        def solve(day, buy):
            
            if(day >= n):
                return 0;
            
            if (day, buy) in dp:
                return dp[(day,buy)]
            
            profit = 0
            if buy:
                take = solve(day+1, not buy) - prices[day]
                not_take = solve(day+1, buy)
                profit = max(profit, take, not_take)
            else:
                sell = prices[day] + solve(day+2, not buy)
                not_sell = solve(day+1, buy)
                profit = max(profit, sell, not_sell)
            
            dp[(day,buy)] = profit
            return dp[(day,buy)]
        
        return solve(0, True)


######################### DP #####################

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if(n==0 or n==1):
            return 0
        dp = [0] * n
        
        dp[0] = 0
        dp[1] = max(prices[1]-prices[0],0)
        
        for i in range(2, len(prices)):
            dp[i] = dp[i-1]
            for j in range(0, i):
                today_profit = prices[i] - prices[j]
                prevpro = dp[j-2] if j>= 2 else 0
                
                dp[i] = max(dp[i], today_profit + prevpro)
        return dp[n-1]

File: C++/Array/309_Best Time to Buy and Sell Stock with Cooldown.cpp
################ Memoziation ##########################
class Solution {
public:
    int t[5001][2];
    int solve(vector<int> prices, int day, int n, bool buy){
        if(day >= n)
            return 0;
        
        if(t[day][buy]!= -1){
            return t[day][buy];
        }
        int profit = 0;
        if(buy){
            int take = solve(prices,day+1, n, false) - prices[day];  // function call for selling then sell - buy = profit
            int nottake = solve(prices,day+1, n, true);
            profit = max({profit, take, nottake});
        }
        else{
            int sell = prices[day] + solve(prices, day+2,n,true);
            int notsell = solve(prices, day+1,n,false);
            profit = max({profit, sell,notsell});
        }
        return t[day][buy] = profit;
    }
    
    
    
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        bool buy = true;
        memset(t,-1,sizeof(t));
        return solve(prices, 0, n,buy);
    }
};


######################### DP #####################

class Solution {
public:  
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n==0 || n==1)
            return 0;
        
        vector<int> t(n,0);
        
        t[0] = 0;
        t[1] = max(prices[1]-prices[0],0);
        
        for(int i = 2; i<n;i++){
            t[i] = t[i-1];
            for(int j = 0; j<=i-1;j++){
                int today_profit = prices[i] - prices[j];
                int prevPro = j >= 2? t[j-2]: 0;
                
                t[i] = max(t[i], today_profit + prevPro);
            }    
        }
            return t[n-1];
        
    }
};